newIDE/app/src/ProjectsStorage/CloudStorageProvider/CloudProjectWriter.js

```diff
@@ -0,0 +1,182 @@
+// @flow
+import * as React from 'react';
+import { type AuthenticatedUser } from '../../Profile/AuthenticatedUserContext';
+import { type FileMetadata } from '..';
+import {
+  commitVersion,
+  createCloudProject,
+  getCredentialsForCloudProject,
+  updateCloudProject,
+} from '../../Utils/GDevelopServices/Project';
+import { serializeToJSON } from '../../Utils/Serializer';
+import { initializeZipJs } from '../../Utils/Zip.js';
+import CloudSaveAsDialog from './CloudSaveAsDialog';
+
+const zipProject = async (project: gdProject) => {
+  const zipJs: ZipJs = await initializeZipJs();
+  const projectJson = serializeToJSON(project);
+  const textReader = new zipJs.TextReader(projectJson);
+
+  return new Promise((resolve, reject) => {
+    try {
+      zipJs.createWriter(new zipJs.BlobWriter('application/zip'), zipWriter => {
+        zipWriter.add('game.json', textReader, () => {
+          zipWriter.close(blob => {
+            resolve(blob);
+          });
+        });
+      });
+    } catch (error) {
+      console.error('An error occurred when zipping project', error);
+      reject(error);
+    }
+  });
+};
+
+const zipProjectAndCommitVersion = async ({
+  authenticatedUser,
+  project,
+  cloudProjectId,
+}: {
+  authenticatedUser: AuthenticatedUser,
+  project: gdProject,
+  cloudProjectId: string,
+}): Promise<?string> => {
+  const archive = await zipProject(project);
+  const newVersion = await commitVersion(
+    authenticatedUser,
+    cloudProjectId,
+    archive
+  );
+  return newVersion;
+};
+
+export const generateOnSaveProject = (
+  authenticatedUser: AuthenticatedUser
+) => async (project: gdProject, fileMetadata: FileMetadata) => {
+  const newVersion = await zipProjectAndCommitVersion({
+    authenticatedUser,
+    project,
+    cloudProjectId: fileMetadata.fileIdentifier,
+  });
+  if (!newVersion) return { wasSaved: false, fileMetadata };
+  return {
+    wasSaved: true,
+    fileMetadata,
+  };
+};
+
+export const generateOnChangeProjectProperty = (
+  authenticatedUser: AuthenticatedUser
+) => async (
+  project: gdProject,
+  fileMetadata: FileMetadata,
+  properties: { name: string }
+): Promise<boolean> => {
+  if (!authenticatedUser.authenticated) return false;
+  try {
+    await updateCloudProject(
+      authenticatedUser,
+      fileMetadata.fileIdentifier,
+      properties
+    );
+    const newVersion = await zipProjectAndCommitVersion({
+      authenticatedUser,
+      project,
+      cloudProjectId: fileMetadata.fileIdentifier,
+    });
+    if (!newVersion) {
+      throw new Error("Couldn't save project following property update.");
+    }
+
+    return true;
+  } catch (error) {
+    // TODO: Determine if a feedback should be given to user so that they can try again if necessary.
```

Comment:
It's a bit difficult here.
They were changing the project properties, they got a warning saying that the project would be saved.
At this point, the properties of the project have been changed but:
- Either the name of the cloud project (in our api) has not been changed (because the error comes from `updateCloudProject`)
- Either the name of the cloud project has been saved but a new version has not been uploaded.
In both cases, when they return to the project properties dialog, they see things just as they think they would be, with no discrepancies.

I was thinking of having a snack that appears in the lower right side where you can hit retry so that the api calls can be sent again. If it succeeds, great, but if not, I'm not sure what to do.

Since this component will be added later, I thought that this TODO could wait, having faith in this api calls.

Reply:
alright 👍 
It could indeed be a small snackbar saying "there was an error while saving your project, hit ctrl+s to try again"